### ISSUE TITLE
Any plans on TypeScript support and d.ts generation?

### ISSUE THREAD

**Issue by stgolem** [dfilatov/vidom#243]
TypeScript is so much cooler than plain JavaScript! And we have typings for everything!
But not for vidom :disappointed:

Have you got plans for it?

**Comment by dfilatov** [comment]
Personally, I'm not a fan of TypeScript, but I agree that typescript definitions could be helpful for people which use it.
If you're familiar with TypeScript, could you add definitions and make pull request? It shouldn't be difficult.

**Comment by stgolem** [comment]
I'm not that good at TypeScript also. Never created definitions for complex modules.
Also i am not that good at vidom (just started to learn and try it). 
So maybe someone else out there in community?

**Comment by dfilatov** [comment]
If you have any questions about Vidom, feel free to ask )

**Comment by stgolem** [comment]
The best way to create typings is to know the structure of classes and objects. If you have something like that, maybe internally, it would help a lot.
From documentation i cannot see much, mostly there are just generic definitions.
What we need is full structure with inheritance and types used in params.

**Comment by dfilatov** [comment]
Let's discuss, for instance for `children()` method of `TagNode` (https://github.com/dfilatov/vidom/wiki/TagNode-API#children)

```
TagNode children(
    VNode child | Array<VNode> children | String text
)
```

Could you write what is not clear here and what you would expect to see to help you?

**Comment by stgolem** [comment]
Ok. Here you go.
What is TagNode? A class? 
Is there any nested classes?
What is TagNode properties (get, set) and functions?
How do i create TagNode?
Same for VNode.

See? You have high level API documented, with examples and all. But we need class structure.
In other case i have to go to sources and spend time composing that.

**Comment by dfilatov** [comment]
Ok, let's start with top-level definitions and creating:

`VNode = TagNode | TextNode | FragmentNode | ComponentNode | FunctionComponentNode`

> Is there any nested classes?

No

> How do i create TagNode?

All types of nodes are created via top-level factory function `node(...)`.

  * `node(string tagName)` creates `TagNode`
  * `node(String "fragment")` creates `FragmentNode` 
  * `node(String "text")` creates `TextNode`
  * `node(Function ComponentClass)` creates `ComponentNode`
  * `node(Function FunctionComponentClass)` creates `FunctionComponentNode`

**Comment by dfilatov** [comment]
Methods of `TagNode`:
  * `key(String key)`
  * `ns(String ns)`
  *  `attrs(Object attrs)`
  * `children(String | VNode | Array VNode children)`
  * `html(String html)`

**Comment by dfilatov** [comment]
Methods of `TextNode`:
  * `key(String key)`
  * `children(String children)`

**Comment by dfilatov** [comment]
Methods of `FragmentNode`:
  * `key(String key)`
  * `children(VNode | Array VNode)`
  * `ctx(Object ctx)`

**Comment by dfilatov** [comment]
Methods of `ComponentNode`:
  * `key(String key)`
  * `attrs(Object attrs)`
  * `children(String | VNode | Array VNode children)`
  * `ctx(Object ctx)`

**Comment by dfilatov** [comment]
@stgolem Is it enough to go ahead?

**Comment by stgolem** [comment]
That is great start!
I think from here and docs i can compose some typings already.
If i have further questions i'll ask here.

**Comment by dfilatov** [comment]
I'm gonna add `vidom.d.ts` in 0.9

**Comment by dfilatov** [comment]
closed in favour of #306
